### PR TITLE
Add HttpResponseBuilder to fluently build `HttpResponse`

### LINKED ISSUE #3398: Provide a way to fluently build an `HttpResponse`
`HttpResponse` only provides various factory methods. It would be convenient to add `HttpResponseBuilder` as like [`HttpRequestBuilder`](https://github.com/line/armeria/blob/461b0fad53e1515387ebbab56a09c4ec0ed20fe6/core/src/main/java/com/linecorp/armeria/common/HttpRequestBuilder.java#L29)
For example:
```java
HttpResponse.builder()
            .status(HttpStatus.OK)
            .content("Hello, Armeria!")
            .contentType(MediaType.PLAIN_TEXT_UTF_8)
            .setCookie(...)
            .header("headerName", "headerValue")
            .build();

Publisher<HttpData> content = ...
HttpResponse.builder()
            .ok()
            .content(content)
            .build(); 
```

### PR DESCRIPTION
**Pull request by ta7uw**
Motivation:

It would be convenient to add `HttpResponseBuilder` as like `HttpRequestBuilder`

Modifications:

- Add `HttpResponseBuilder` to build `HttpResponse`
- Add `HttpResponse#builder` to  get a builder instance

Result:

- Closes #3398
-  Fluently build `HttpResponse` by using this builder as like HttpRequestBuilder
```
HttpResponse.builder()
            .ok()
            .content(content)
            .build(); 
```

### CODE REVIEW

File: core/src/main/java/com/linecorp/armeria/common/HttpResponseBuilder.java [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+    }
+
+    /**
+     * Sets the content for this response.
+     */
+    public HttpResponseBuilder content(MediaType contentType, String content) {
+        requireNonNull(contentType, "contentType");
+        requireNonNull(content, "content");
+        return content(contentType, HttpData.of(contentType.charset(StandardCharsets.UTF_8), content));
+    }
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/common/HttpResponseBuilder.java:152]: We may extract the common part between `AbstractHttpRequestBuilder` and `HttpResponseBuilder` to `AbstractHttpMessageBuilder`.

Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/common/HttpResponseBuilder.java:152]: Would be useful if we add `contentJson(Object content)`?

Comment (ta7uw) [reply core/src/main/java/com/linecorp/armeria/common/HttpResponseBuilder.java:152]: > We may extract the common part between AbstractHttpRequestBuilder and HttpResponseBuilder to AbstractHttpMessageBuilder.

Should I address it in this PR?

> Would be useful if we add contentJson(Object content)?

Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/common/HttpResponseBuilder.java:152]: How about also adding the `content` method that takes `Publisher` as an argument?
```
public HttpResponseBuilder content(Publisher<? extends HttpData> content) {...}
```

Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/common/HttpResponseBuilder.java:152]: > Should I address it in this PR?

I hope so. 😀 Because duplicate code is caused by this new feature.

Comment (ta7uw) [reply core/src/main/java/com/linecorp/armeria/common/HttpResponseBuilder.java:152]: > How about also adding the content method that takes Publisher as an argument?

It seems to be nice.

File: core/src/main/java/com/linecorp/armeria/common/AbstractHttpMessageBuilder.java [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+    public abstract AbstractHttpMessageBuilder content(MediaType contentType,
+                                                       Publisher<? extends HttpData> content);
+
+    /**
+     * Sets the content for this response. The {@code content} that is converted into JSON
+     * using the default {@link ObjectMapper}.
+     */
+    public AbstractHttpMessageBuilder contentJson(Object content) {
+        try {
+            this.content = HttpData.wrap(JacksonUtil.writeValueAsBytes(content));
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/common/AbstractHttpMessageBuilder.java:124]: We need to set `application/json` as a content type for JSON format.

File: core/src/main/java/com/linecorp/armeria/common/AbstractHttpMessageBuilder.java [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+
+    @Nullable
+    protected HttpData getContent() {
+        return content;
+    }
+
+    @Nullable
+    protected Publisher<? extends HttpData> getPublisher() {
+        return publisher;
+    }
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/common/AbstractHttpMessageBuilder.java:50]: We usually don't use `get` prefix for getters.
```suggestion
    @Nullable
    protected final HttpData content() {
        return content;
    }

    @Nullable
    protected final Publisher<? extends HttpData> pblisher() {
        return publisher;
    }
```

File: core/src/main/java/com/linecorp/armeria/common/AbstractHttpMessageBuilder.java [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+
+    /**
+     * Sets the {@link Publisher} for this response.
+     */
+    public AbstractHttpMessageBuilder content(Publisher<? extends HttpData> content) {
+        requireNonNull(content, "publisher");
+        checkState(this.content == null, "content has been set already");
+        publisher = content;
+        return this;
+    }
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/common/AbstractHttpMessageBuilder.java:69]: As `HttpData` is binary data, we can't guess the `MediaType`. could we remove these two builder methods? I guess that an implicit `MediaType` can be confusing to users.

File: core/src/main/java/com/linecorp/armeria/common/AbstractHttpMessageBuilder.java [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+    /**
+     * Sets the content for this response.
+     */
+    public abstract AbstractHttpMessageBuilder content(MediaType contentType, HttpData content);
+
+    /**
+     * Sets the {@link Publisher} for this response.
+     */
+    public abstract AbstractHttpMessageBuilder content(MediaType contentType,
+                                                       Publisher<? extends HttpData> content);
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/common/AbstractHttpMessageBuilder.java:116]: I think we can remove `abstract` from some methods and delegate to other methods that could simplify the implementation of subclasses.
For example:
```java
public AbstractHttpMessageBuilder content(MediaType contentType, byte[] content) {
    return content(contentType, HttpData.wrap(content));
}
```

File: core/src/main/java/com/linecorp/armeria/common/AbstractHttpMessageBuilder.java (hunk 4 above)
Comment (ta7uw) [reply core/src/main/java/com/linecorp/armeria/common/AbstractHttpMessageBuilder.java:69]: ok, remove them.

File: core/src/main/java/com/linecorp/armeria/common/AbstractHttpRequestBuilder.java [hunk 6]
```diff
@@ -391,6 +402,8 @@ public AbstractHttpRequestBuilder cookies(Iterable<? extends Cookie> cookies) {
      */
     protected final HttpRequest buildRequest() {
         final RequestHeaders requestHeaders = requestHeaders();
+        final HttpData content = content();
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/common/AbstractHttpRequestBuilder.java:405]: Let's do this in line 415. 😉

File: core/src/main/java/com/linecorp/armeria/common/AbstractHttpRequestBuilder.java [hunk 7]
```diff
@@ -416,9 +429,14 @@ protected final HttpRequest buildRequest() {
...
+        final MediaType contentType = contentType();
+        if (contentType != null) {
+            requestHeadersBuilder.contentType(contentType);
+        }
         if (cookies != null) {
             requestHeadersBuilder.set(COOKIE, Cookie.toCookieHeader(cookies));
         }
+        final HttpData content = content();
         if (content == null || content.isEmpty()) {
             requestHeadersBuilder.remove(CONTENT_LENGTH);
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/common/AbstractHttpRequestBuilder.java:441]: A user can set the content length with a `Publisher` based on the pre-calculation. 
So, I think we shouldn't remove it when `Publisher` is not null.

File: core/src/main/java/com/linecorp/armeria/common/HttpResponseBuilder.java [hunk 8]
```diff
@@ -0,0 +1,270 @@
...
+    public HttpResponseBuilder headers(Iterable<? extends Entry<? extends CharSequence, String>> headers) {
+        requireNonNull(headers, "headers");
+        responseHeadersBuilder.set(headers);
+        return this;
+    }
+
+    /**
+     * Sets HTTP trailers for this response.
+     */
+    public HttpResponseBuilder trailers(Iterable<? extends Entry<? extends CharSequence, String>> trailers) {
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/common/HttpResponseBuilder.java:243]: How about moving this up to `AbstractHttpMessageBuilder` because an `HttpRequest` can have a trailers?
https://datatracker.ietf.org/doc/html/rfc7540#section-8.1

**Review by minwoox — COMMENTED** [review]
Basically, looks good! 😄

File: core/src/main/java/com/linecorp/armeria/common/HttpResponseBuilder.java [hunk 9]
```diff
@@ -0,0 +1,270 @@
...
+     * <pre>{@code
+     * HttpResponse.builder()
+     *             .ok()
+     *             .content("Hello, Armeria")
+     *             .headers(HttpHeaders.of("x-test-header", "foo", "Server", "baz"))
+     *             .build();
+     * }</pre>
+     * @see HttpHeaders
+     */
+    public HttpResponseBuilder headers(Iterable<? extends Entry<? extends CharSequence, String>> headers) {
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/common/HttpResponseBuilder.java:234]: Let's make an abstract API for this in `AbstractHttpMessageBuilder`.

File: core/src/main/java/com/linecorp/armeria/common/HttpResponseBuilder.java [hunk 10]
```diff
@@ -0,0 +1,270 @@
...
+        requireNonNull(headers, "headers");
+        responseHeadersBuilder.set(headers);
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/common/HttpResponseBuilder.java:236]: I think it is a bug to use `set` here. https://github.com/line/armeria/issues/3932#issuecomment-970254885
Could you use `add` for `header` and `headers` instead.

File: core/src/main/java/com/linecorp/armeria/common/HttpResponseBuilder.java [hunk 11]
```diff
@@ -0,0 +1,270 @@
...
+
+/**
+ * Builds a new {@link HttpResponse}.
+ */
+public final class HttpResponseBuilder extends AbstractHttpMessageBuilder {
+
+    private final ResponseHeadersBuilder responseHeadersBuilder = ResponseHeaders.builder();
+
+    private final HttpHeadersBuilder httpTrailers = HttpHeaders.builder();
+
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/common/HttpResponseBuilder.java:38]: Could you add a package-private constructor to prevent creating directly?
```suggestion
    HttpResponseBuilder() {}
```

File: core/src/main/java/com/linecorp/armeria/common/HttpResponseBuilder.java (hunk 8 above)
Comment (ta7uw) [reply core/src/main/java/com/linecorp/armeria/common/HttpResponseBuilder.java:243]: ok, I move trailers to the abstract class.

File: core/src/main/java/com/linecorp/armeria/common/HttpResponseBuilder.java [hunk 12]
```diff
@@ -0,0 +1,270 @@
...
+            if (content == null) {
+                content = HttpData.empty();
+            }
+            if (trailers == null) {
+                return HttpResponse.of(responseHeaders, content);
+            } else {
+                return HttpResponse.of(responseHeaders, content, trailers.build());
+            }
+        } else {
+            return HttpResponse.of(responseHeaders, publisher);
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/common/HttpResponseBuilder.java:267]: I think we should add trailers if it's not null as we do in the request builder.

**Review by minwoox — COMMENTED** [review]
Thanks! I think we are almost there. 😄

File: core/src/main/java/com/linecorp/armeria/common/AbstractHttpMessageBuilder.java [hunk 13]
```diff
@@ -0,0 +1,152 @@
...
+import com.google.errorprone.annotations.FormatMethod;
+import com.google.errorprone.annotations.FormatString;
+
+import com.linecorp.armeria.common.annotation.Nullable;
+import com.linecorp.armeria.internal.common.JacksonUtil;
+
+@SuppressWarnings("checkstyle:OverloadMethodsDeclarationOrder")
+abstract class AbstractHttpMessageBuilder {
+
+    private final HttpHeadersBuilder httpHeaders = HttpHeaders.builder();
```
[the remark this answers is not in the document]
Comment (ta7uw) [reply core/src/main/java/com/linecorp/armeria/common/AbstractHttpMessageBuilder.java:37]: It seems good to me.

File: core/src/main/java/com/linecorp/armeria/common/AbstractHttpRequestBuilder.java [hunk 14]
```diff
@@ -399,6 +405,7 @@ protected final HttpRequest buildRequest() {
                                       StreamMessage.concat(publisher, StreamMessage.of(httpTrailers.build())));
```
[the remark this answers is not in the document]
Comment (ta7uw) [reply core/src/main/java/com/linecorp/armeria/common/AbstractHttpRequestBuilder.java:405]: It seems useful!
I will work on it.

File: core/src/main/java/com/linecorp/armeria/common/AbstractHttpMessageBuilder.java [hunk 15]
```diff
@@ -0,0 +1,148 @@
...
+        return this;
+    }
+
+    AbstractHttpMessageBuilder contentJson(Object content) {
+        requireNonNull(content, "content");
+        headersBuilder().contentType(MediaType.JSON);
+        try {
+            this.content = HttpData.wrap(JacksonUtil.writeValueAsBytes(content));
+        } catch (JsonProcessingException e) {
+            throw new IllegalArgumentException(e.toString(), e);
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/common/AbstractHttpMessageBuilder.java:134]: nit:
```suggestion
            throw new IllegalArgumentException("failed to serialize " + content, e);
```
Let's fix it here as well. https://github.com/line/armeria/blob/master/core/src/main/java/com/linecorp/armeria/common/HttpResponse.java#L558

File: core/src/main/java/com/linecorp/armeria/common/HttpResponseBuilder.java [hunk 16]
```diff
@@ -0,0 +1,281 @@
...
+     *             .content("Hello, Armeria")
+     *             .header("Server", "foo")
+     *             .build();
+     * }</pre>
+     */
+    @Override
+    public HttpResponseBuilder header(CharSequence name, Object value) {
+        responseHeadersBuilder.addObject(requireNonNull(name, "name"),
+                                         requireNonNull(value, "value"));
+        return (HttpResponseBuilder) super.header(name, value);
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/common/HttpResponseBuilder.java:223]:
```suggestion
        return (HttpResponseBuilder) super.header(name, value);
```
We shouldn't add the header here. 😉

File: core/src/test/java/com/linecorp/armeria/common/HttpResponseBuilderTest.java [hunk 17]
```diff
@@ -0,0 +1,285 @@
...
+                                                      StreamMessage.of(
+                                                              HttpData.ofUtf8(
+                                                                      "Armeriaはいろんな使い方がアルメリア"
+                                                              )
+                                                      ))
+                                             .build();
+        final AggregatedHttpResponse aggregatedRes = res.aggregate().join();
+        assertThat(aggregatedRes.status()).isEqualTo(HttpStatus.OK);
+        assertThat(aggregatedRes.contentUtf8()).isEqualTo("Armeriaはいろんな使い方がアルメリア");
+        assertThat(aggregatedRes.contentType()).isNotNull();
```
Comment (minwoox) [inline core/src/test/java/com/linecorp/armeria/common/HttpResponseBuilderTest.java:182]: We can remove all these checks because we are checking the contentType anyway. 😄

**Review by minwoox — COMMENTED** [review]
I guess this is the final round. 😄

File: core/src/main/java/com/linecorp/armeria/common/HttpResponseBuilder.java (hunk 16 above)
Comment (ta7uw) [reply core/src/main/java/com/linecorp/armeria/common/HttpResponseBuilder.java:223]: Oh, fix it

**Comment by ikhoon** [comment]
Added `defect` label because this PR also fixes https://github.com/line/armeria/issues/3932#issuecomment-970254885

File: core/src/main/java/com/linecorp/armeria/common/HttpResponseBuilder.java [hunk 18]
```diff
@@ -0,0 +1,279 @@
...
+     */
+    @Override
+    public HttpResponseBuilder trailers(Iterable<? extends Entry<? extends CharSequence, String>> trailers) {
+        return (HttpResponseBuilder) super.trailers(trailers);
+    }
+
+    /**
+     * Builds the response.
+     */
+    public HttpResponse build() {
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/common/HttpResponseBuilder.java:251]: By convention, could you place this method at the end of this class?

File: core/src/test/java/com/linecorp/armeria/common/HttpResponseBuilderTest.java [hunk 19]
```diff
@@ -0,0 +1,277 @@
...
+    }
+
+    @Test
+    void buildWithHeaders() {
+        final HttpResponse res = HttpResponse.builder()
+                                             .ok()
+                                             .headers(HttpHeaders.of("header-1",
+                                                                     "test-value1",
+                                                                     "header-2",
+                                                                     "test-value2"))
```
Comment (ikhoon) [inline core/src/test/java/com/linecorp/armeria/common/HttpResponseBuilderTest.java:212]: Could you add a test that adds duplicate header names that was issued in #3932

**Review by ikhoon — COMMENTED** [review]
Almost there! Left only minor nits.

File: core/src/main/java/com/linecorp/armeria/common/AbstractHttpMessageBuilder.java [hunk 20]
```diff
@@ -0,0 +1,148 @@
...
+    }
+
+    AbstractHttpMessageBuilder headers(
+            Iterable<? extends Entry<? extends CharSequence, String>> headers) {
+        requireNonNull(headers, "headers");
+        headersBuilder().add(headers);
+        return this;
+    }
+
+    AbstractHttpMessageBuilder content(String content) {
```
[the remark this answers is not in the document]
Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/common/AbstractHttpMessageBuilder.java:76]: I didn't request it because Javadoc of subclasses has request or response in the sentence depending on its type. But either is fine to me. 😉

[the remark this answers is not in the document]
Comment (ta7uw) [reply core/src/main/java/com/linecorp/armeria/common/AbstractHttpMessageBuilder.java:76]: > I didn't request it because Javadoc of subclasses has request or response in the sentence depending on its type.

The reason why I wrote Javadoc in subclasses is this.

File: core/src/test/java/com/linecorp/armeria/common/HttpResponseBuilderTest.java [hunk 21]
```diff
@@ -0,0 +1,284 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.common;
+
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+import static org.assertj.core.api.AssertionsForClassTypes.assertThat;
```
Comment (jrhee17) [inline core/src/test/java/com/linecorp/armeria/common/HttpResponseBuilderTest.java:20]: Prefer to use `import static org.assertj.core.api.Assertions.assertThat;` so that we can take full advantage of the `assertj` API

File: core/src/test/java/com/linecorp/armeria/common/HttpResponseBuilderTest.java [hunk 22]
```diff
@@ -0,0 +1,284 @@
...
+                                                                      "trailer-value"))
+                                             .build();
+        final AggregatedHttpResponse aggregatedRes = res.aggregate().join();
+        assertThat(aggregatedRes.status()).isEqualTo(HttpStatus.OK);
+        assertThat(aggregatedRes.headers().contains("header-1")).isTrue();
+        assertThat(aggregatedRes.headers().contains("header-2")).isTrue();
+        assertThat(aggregatedRes.headers().get("header-1")).isEqualTo("test-value1");
+        assertThat(aggregatedRes.headers().get("header-2")).isEqualTo("test-value2");
+        assertThat(aggregatedRes.headers().getAll(HttpHeaderNames.ACCEPT_ENCODING))
+                .isEqualTo(ImmutableList.of("gzip", "deflate", "gzip"));
```
Comment (jrhee17) [inline core/src/test/java/com/linecorp/armeria/common/HttpResponseBuilderTest.java:259]: nit; prefer the assertj api for richer test failure messages (and better style!)

```suggestion
        assertThat(aggregatedRes.headers().getAll(HttpHeaderNames.ACCEPT_ENCODING))
                .containsExactly("gzip", "deflate", "gzip");
```

**Review by jrhee17 — APPROVED** [review]
The changes look really nice 👍  LGTM apart from minor comments. Thank you! 🙇

File: core/src/main/java/com/linecorp/armeria/common/AbstractHttpMessageBuilder.java [hunk 23]
```diff
@@ -0,0 +1,148 @@
...
+        return content(MediaType.PLAIN_TEXT_UTF_8, content);
+    }
+
+    AbstractHttpMessageBuilder content(MediaType contentType, CharSequence content) {
+        requireNonNull(contentType, "contentType");
```
[the remark this answers is not in the document]
Comment (ta7uw) [reply core/src/main/java/com/linecorp/armeria/common/AbstractHttpMessageBuilder.java:81]: It seems good.

File: core/src/main/java/com/linecorp/armeria/common/AbstractHttpMessageBuilder.java (hunk 20 above)
[the remark this answers is not in the document]
Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/common/AbstractHttpMessageBuilder.java:76]: OK to leave it as it is.

File: core/src/main/java/com/linecorp/armeria/common/AbstractHttpRequestBuilder.java [hunk 24]
```diff
@@ -414,13 +421,21 @@
...
+
     private RequestHeaders requestHeaders() {
         requestHeadersBuilder.path(buildPath());
         if (cookies != null) {
             requestHeadersBuilder.set(COOKIE, Cookie.toCookieHeader(cookies));
         }
+        final HttpData content = content();
         if (content == null || content.isEmpty()) {
-            requestHeadersBuilder.remove(CONTENT_LENGTH);
+            if (publisher() == null) {
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/common/AbstractHttpRequestBuilder.java:436]: Revert? A publisher will be sent with chunked transfer encoding.

File: core/src/main/java/com/linecorp/armeria/common/HttpResponseBuilder.java [hunk 25]
```diff
@@ -0,0 +1,270 @@
...
+     * Sets the content for this response. The {@code content} is converted into JSON format
+     * using the default {@link ObjectMapper}.
+     */
+    @Override
+    public HttpResponseBuilder contentJson(Object content) {
+        return (HttpResponseBuilder) super.contentJson(content);
+    }
+
+    /**
+     * Sets a header for this response. For example:
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/common/HttpResponseBuilder.java:201]: `Adds`

File: core/src/main/java/com/linecorp/armeria/common/HttpResponseBuilder.java [hunk 26]
```diff
@@ -0,0 +1,270 @@
...
+     *             .build();
+     * }</pre>
+     */
+    @Override
+    public HttpResponseBuilder header(CharSequence name, Object value) {
+        return (HttpResponseBuilder) super.header(name, value);
+    }
+
+    /**
+     * Sets multiple headers for this response. For example:
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/common/HttpResponseBuilder.java:216]: `Adds`

**Review by ikhoon — COMMENTED** [review]
Looking good. Left only minor nits.

File: core/src/main/java/com/linecorp/armeria/common/AbstractHttpMessageBuilder.java [hunk 27]
```diff
@@ -0,0 +1,153 @@
...
+        checkState(publisher == null, "publisher has been set already");
+        headersBuilder().contentType(contentType);
+        this.content = content;
+        return this;
+    }
+
+    AbstractHttpMessageBuilder content(MediaType contentType, Publisher<? extends HttpData> publisher) {
+        requireNonNull(contentType, "contentType");
+        requireNonNull(publisher, "publisher");
+        checkState(this.content == null, "content has been set already");
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/common/AbstractHttpMessageBuilder.java:127]: nit:
```suggestion
        checkState(content == null, "content has been set already");
```

**Review by minwoox — APPROVED** [review]
Great job! Thanks, @ta7uw! 😄

**Comment by ikhoon** [comment]
> Revert? A publisher will be sent with chunked transfer encoding.

Had a chat internally, my previous comment does not comply with Armeria's policy.
If users can predict the total size of a `Publisher`, we need to respect it.
Please revive `if (publisher() == null) { ` condition.
Sorry for the back and forth. 😅

File: core/src/main/java/com/linecorp/armeria/common/AbstractHttpRequestBuilder.java [hunk 28]
```diff
@@ -399,6 +405,7 @@ protected final HttpRequest buildRequest() {
                                       StreamMessage.concat(publisher, StreamMessage.of(httpTrailers.build())));
             }
         }
+        final HttpData content = content();
```
[the remark this answers is not in the document]
Comment (ta7uw) [reply core/src/main/java/com/linecorp/armeria/common/AbstractHttpRequestBuilder.java:408]: OK

File: core/src/main/java/com/linecorp/armeria/common/HttpResponseBuilder.java [hunk 29]
```diff
@@ -0,0 +1,270 @@
...
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.google.errorprone.annotations.FormatMethod;
+import com.google.errorprone.annotations.FormatString;
+
+import com.linecorp.armeria.common.stream.StreamMessage;
+
+/**
+ * Builds a new {@link HttpResponse}.
+ */
+public final class HttpResponseBuilder extends AbstractHttpMessageBuilder {
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/common/HttpResponseBuilder.java:35]: Could you add `@UnstableApi`?

File: core/src/main/java/com/linecorp/armeria/common/HttpResponse.java [hunk 30]
```diff
@@ -617,6 +617,13 @@ static HttpResponse ofFailure(Throwable cause) {
         return new AbortedHttpResponse(cause);
     }
 
+    /**
+     * Returns a new {@link HttpResponseBuilder}.
+     */
+    static HttpResponseBuilder builder() {
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/common/HttpResponse.java:623]: Could you add `@UnstableApi`?

**Review by ikhoon — APPROVED** [review]
Thanks a lot, @ta7uw! 🎉🚀
Left a minor comment. 🙇‍♂️

**Comment by minwoox** [comment]
Now I can fluently build my response. 🤣 
Great work, @ta7uw! 🎉 🎉 🎉 
And thanks, reviewers!
